### PR TITLE
*: Collation should be set accordingly when setting string datum

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

For collation support.

### What is changed and how it works?

This pr only change the method `SetString` of datum so it can store the collation information.

### Check List 

Tests 

 - Existing Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -61,7 +62,7 @@ const (
 // It has better performance and is easier to use than `interface{}`.
 type Datum struct {
 	k         byte        // datum kind.
-	collation uint8       // collation can hold uint8 values.
+	collation string      // collation hold the collation information for string value.
```
Comment (bb7133) [inline types/datum.go]: I still prefer using an integer(maybe `uint16`?) for collation, and we can use `GetCollatorByID()` to get them. After all, integer use much more memory than strings.

File: executor/load_data.go [hunk 2]
```diff
@@ -473,7 +474,7 @@ func (e *LoadDataInfo) colsToRow(ctx context.Context, cols []field) []types.Datu
 		if cols[i].maybeNull && string(cols[i].str) == "N" {
 			e.row[i].SetNull()
 		} else {
-			e.row[i].SetString(string(cols[i].str))
+			e.row[i].SetString(string(cols[i].str), collate.DefaultCollation, collate.DefaultLen)
```
Comment (wjhuang2016) [inline executor/load_data.go]: Why use default？

File: expression/distsql_builtin.go [hunk 3]
```diff
@@ -1169,9 +1172,10 @@ func convertUint(val []byte) (*Constant, error) {
 	return &Constant{Value: d, RetType: &types.FieldType{Tp: mysql.TypeLonglong, Flag: mysql.UnsignedFlag}}, nil
 }
 
-func convertString(val []byte) (*Constant, error) {
+func convertString(val []byte, tp *tipb.FieldType) (*Constant, error) {
+	log.Warn("????", zap.String("tp", fmt.Sprintf("%v", tp)))
```
Comment (wjhuang2016) [inline expression/distsql_builtin.go]: ？

File: executor/load_data.go (hunk 2 above)
Comment (winoros) [reply executor/load_data.go]: Here, haven't touch the column information. So no column filed type information can be got.

File: expression/distsql_builtin.go (hunk 3 above)
Comment (bb7133) [reply expression/distsql_builtin.go]: Please update this warning message.

File: types/datum.go [hunk 4]
```diff
@@ -1692,7 +1692,13 @@ func NewBytesDatum(b []byte) (d Datum) {
 
 // NewStringDatum creates a new Datum from a string.
 func NewStringDatum(s string) (d Datum) {
```
Comment (qw4990) [inline types/datum.go]: Should we add arguments `collation` and `len` for it?

Comment (qw4990) [inline types/datum.go]: Since we have `NewCollationStringDatum` now, should we remove this function?

File: util/collate/collate.go [hunk 5]
```diff
@@ -139,6 +157,7 @@ func (bpc *binPaddingCollator) Key(str string, opt CollatorOption) []byte {
 func init() {
 	collatorMap = make(map[string]Collator)
 	collatorIDMap = make(map[int]Collator)
+	id2NameMap = make(map[int32]string)
```
Comment (bb7133) [inline util/collate/collate.go]: We have the same defined in `parser` already:
https://github.com/pingcap/parser/blob/517beb2e39c20e78cca59b63351aacb5c1ef5845/mysql/charset.go#L126

Please use this one instead.

File: util/collate/collate.go [hunk 6]
```diff
@@ -21,10 +21,18 @@ import (
...
 	collatorIDMap       map[int]Collator
+	id2NameMap          map[int32]string
 	newCollationEnabled bool
 	setCollationOnce    sync.Once
 )
 
+// DefaultCollation is default collation of TiDB. If a string column/constant is defined without collation,
+// we use this as its collation.
+const (
+	DefaultCollation = "binary"
```
Comment (bb7133) [inline util/collate/collate.go]: Please notice that the default collation of string in TiDB is not `binary`, it is `utf8mb4_bin`:

https://github.com/pingcap/parser/blob/517beb2e39c20e78cca59b63351aacb5c1ef5845/mysql/charset.go#L587

**Pull request merged**